### PR TITLE
logictest: don't download `cockroach` from the internet in test

### LINKED ISSUE #108865: logictest: `cockroach-go-testserver-upgrade-to-master` should not hit the network
These tests download `cockroach` from binaries.cockroachdb.com. Our unit tests shouldn't download anything from the internet.

Epic CRDB-8308

Jira issue: CRDB-30671

### PR DESCRIPTION
**Pull request by rickystewart**
Instead, make it a Bazel-level dependency and look it up in runfiles. This means the test no longer has to talk to the network.

Closes #108865.

Epic: CRDB-8308
Release note: None

### CODE REVIEW

File: pkg/sql/logictest/REPOSITORIES.bzl [hunk 1]
```diff
@@ -0,0 +1,46 @@
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# When updating `cockroach_releases.yaml`, update the _PREDECESSOR_VERSION
+# and the SHA256 sums in _CONFIGS.
```
Comment (rail) [inline pkg/sql/logictest/REPOSITORIES.bzl:4]: This requires an extra step in the release process or in https://github.com/cockroachdb/cockroach/blob/master/.github/workflows/update_releases.yaml. Should we create a template for this file or change the values in place? 🤔 
What happens if we forget updating these?

Comment (rickystewart) [reply pkg/sql/logictest/REPOSITORIES.bzl:4]: The test will fail if `cockroach_releases.yaml` and this file are not in-sync.

Not sure what you mean by creating a template for this file. My thinking was the person doing the update would update these values in place.

Comment (rickystewart) [reply pkg/sql/logictest/REPOSITORIES.bzl:4]: If you want I can extract this stuff (the SHA's and version number) into its own file without all the other stuff. Does that make things easier?

Comment (renatolabs) [reply pkg/sql/logictest/REPOSITORIES.bzl:4]: > My thinking was the person doing the update would update these values in place.

I don't think people are manually updating `cockroach_releases.yaml`; the GitHub worklow mentioned by Rail is updating it automatically and opening PRs when there's a change.

Could we update [update_releases.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/release/update_releases.go) to automate this step as well?

Comment (rickystewart) [reply pkg/sql/logictest/REPOSITORIES.bzl]: Done. I'll need someone to pay extra attention to `update_releases.go` to make sure I'm doing it right -- I'm not 100% sure I've implemented the logic correctly, though it seems to work for `master`.

**Review by rail — APPROVED** [review]

File: pkg/cmd/release/update_releases.go [hunk 2]
```diff
@@ -80,10 +82,23 @@ func updateReleasesFile(_ *cobra.Command, _ []string) (retErr error) {
...
+	currentVersion := version.MustParse(build.BinaryVersion())
+	addCurrentRelease(result, currentVersion)
 
 	fmt.Printf("writing results to %s\n", releaseDataFile)
-	if err := saveResults(result); err != nil {
+	if err := saveResultsInYaml(result); err != nil {
+		return err
+	}
+	key := fmt.Sprintf("%d.%d", currentVersion.Major(), currentVersion.Minor())
+	predecessor := result[key].Latest
```
Comment (renatolabs) [inline pkg/cmd/release/update_releases.go:93]: Nit: `currentSeries` is probably a better name for `key`.

Shouldn't this always be `result[result[key].Predecessor].Latest`?

For example, when we release the first rc or beta for 23.2, the `23.2` entry in the releases YAML will have a `Latest` field. However, if the `version.txt` file still indicates that master is `23.2`, we should still be using the latest 23.1 as predecessor. We should only start using a 23.2 release as predecessor once `version.txt` (and hence `currentVersion`) is bumped.

**Review by renatolabs — APPROVED** [review]
One question but otherwise LGTM, thanks for automating this!

File: pkg/cmd/release/update_releases.go (hunk 2 above)
Comment (rickystewart) [reply pkg/cmd/release/update_releases.go:93]: > Shouldn't this always be result[result[key].Predecessor].Latest?

I have no idea. If you believe this is what it should be then I'm inclined to think you're right.

**Comment by rickystewart** [comment]
TFTRs!

bors r=rail,renatolabs

**Pull request merged**